src/SGReplicator.cpp

```diff
@@ -68,13 +68,31 @@ namespace Strata {
 
     void SGReplicator::stop() {
         lock_guard<mutex> lock(replicator_lock_);
-        if(c4replicator_ != nullptr){
+        if(c4replicator_ != nullptr) {
             internal_status_ = Strata::SGReplicatorInternalStatus::kStopping;
             c4repl_stop(c4replicator_);
         }
         replicator_can_restart_ = false;
     }
 
+    void SGReplicator::join() {
+        // first wait for the replicator to terminate
+        {
+            lock_guard<mutex> lock(replicator_lock_);
+            if(c4replicator_ != nullptr) {
+                while (c4repl_getStatus(c4replicator_).level != kC4Stopped)
+                    this_thread::sleep_for(chrono::milliseconds(1));
+            }
+        }
+
+        // then wait for the onStatusChanged event to be emitted and terminated
+        unsigned count = 0;     // in case the C4Replicator would fail to emit onStatusChanged event
+        while ((internal_status_ != Strata::SGReplicatorInternalStatus::kStopped) && (count < 200)) {
+            this_thread::sleep_for(chrono::milliseconds(1));
```

Comment:
Sleeping for only 1 ms might be too short - consider sleeping for longer and reducing the max count

Reply:
I checked and it usually takes around ~1 msec to complete the operations under normal scenarios.
Should I change sleep to 5 msec and max count to for example 40?

Reply:
In this case since you checked/timed the operation, I believe this might be fine then. I'd say it's up to you.